src/sas/qtgui/Perspectives/Invariant/InvariantPerspective.py

```diff
@@ -318,6 +318,9 @@ def calculateThread(self, extrapolation):
             try:
                 surface, surface_error = \
                     inv.get_surface_with_error(self._contrast, self._porod)
+                if surface_error == 0:
```

Comment:
`surface_error` is a float. Comparing floats is a tricky thing, maybe we should do the comparison against some very small value?
`if abs(surface_error) < 0.00001` or something like `if math.isclose(surface_error, 0)`

Reply:
The test against zero happens to not a problem in this case because the `get_surface_with_error` function *always* returns zero for the surface error. If it were actually calculated, you probably want to use `surface_error < 1e-7*surface` or some such rather than an absolute test so you don't have to worry about the units used for the surface calculation.

There are bigger problems, though.  The setting of the `surface_error` field is overridden on [line 417](https://github.com/SasView/sasview/blob/af992c426120c8ba6dbeee85b6cac535af375961/src/sas/qtgui/Perspectives/Invariant/InvariantPerspective.py#L417-L420), at least for the case when `bool(surface) is True`.  That means that when surface is non-zero, then a zero will be entered into the surface error field overriding the setting "NONE" being set here.

That `if surface:` test also means that when the `get_surface_with_error` function returns zero the value in the surface field will not be updated, even if it is currently not zero.

A further problem is if an exception is raised here then `surface` is undefined, and the `if surface:` test will throw an exception.

Still another question/concern is that this code is calling `setItem` directly, whereas line 417 is using `reactor.callFromThread` to call `setitem` — why the difference?

So the conditions you want seems to be:
```python
if surface present:
    if get_surface_with_error throws an exception:
        display error in surface and surface error fields
    else:
        display surface value in surface field
        if surface error is zero:
           display none in surface error field
        else:
           display surface error value in surface error field
else:
    display none in surface and surface error fields
```

Maybe hijack `updateModelFromThread` with:
```python
    if value is None:
        item = QtGui.QStandardItem("NONE")
    elif np.isnan(value):
        item = QtGui.QStandardItem("ERROR")
    else:
        item = QtGui.QStandardItem(str(float('%.3g' % value)))
    self.model.setItem(widget, item)
```

Then your logic here would be:
```python
if self._porod:
    try:
        surface, surface_error = ...
        if surface_error < 1e-7*surface:
            surface_error = None
    except Exception as ex:
        calculation_failed = True
        msg += str(ex)
        surface = surface_error = np.NaN
else:
    surface = surface_error = None
```
and below call `updateModelFromThread` for surface and surface_error without the `if surface:` test.

You can do something similar for the qstar logic.

Reply:
Re "... the if surface: test will throw an exception.":  This is not an issue because line 335 checks for `calculation_failed`.

You still may want to use `reactor.callFromThread` to update the fields within the exception code, or better yet, use `calculation_failed` condition to set the extrapolation values to `None` so that the form update logic is done together as a block at the bottom.

Reply:
Thanks guys. This was in fact the one bit of code I was hoping to get feedback on as there was no copy paste from 4.x possible here and while we dealt with in wx, this code is beyond my monkey see monkey do approach (having not seen before 😃  so what we are trying to do is this:  currently the calculation was disabled because it was totally bogus and moreover there is no information with which to compute the uncertainty.  That should be another ticket (I need to check).  The ideas was that if not calculated the GUI should show "none" whereas if it is calculated (at some point in the future when the GUI allows the necessary input AND the user chooses to provide it), no matter how small it is, it should show the numeric value.

Is what you suggest above .. which I've only partially understood so far I'll admit (not had time to really think about it yet) or am I totally barking up the wrong tree here?

Reply:
If you assign zero to a value you can reliably test whether it is zero, but you won't be able to distinguish that from something that is accidentally zero because of floating point math.  

If it is possible that the calculated value is zero and you want to display 0.0, then you should return something else (None?) from the sascalc function when the value is not calculated.

Re `reactor.callFromThread`: I do not know if it is required. In wx you need to use the same thread for all GUI operations or weird things can happen, but I don't know what happens in qt. Maybe @rozyczko can answer?

Reply:
but if the return value has to be a float it cannot be "None" right? maybe nan? but that suggest and error. Or are you allowed to return a "None" object instead of a float? I guess I can try..

Reply:
You can return anything you want . All Python values are objects that carry their type along with them, even floats, and python does nothing to enforce the type of the argument received or the returned value.  The caller has to sort it out by using code like `if value is None:`.

We can hint what we think the type should be, which is handy for the maintainer and can let automated tools (e.g., *mypy*) check that you are using types correctly.  In that case we sould define the function as follows:
```python
from typing import Optional, Tuple

def get_surface_with_error(
    self, 
    contrast: float, 
    porod_const: float, 
    extrapolation: Optional[str]=None,
) -> Tuple[float, Optional[float]]:
    ...
```
or in python2.7 style as used by sasmodels:
```python
def get_surface_with_error(self, contrast, porod_const, extrapolation):
    # type: (float, float, Optional[str]) -> Tuple[float, Optional[float]]
    ...
```